util/reduce_flash_size.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""Script used to reduce the size of the embedded flash for smaller FPGA devices.
+
+This script modifies two source files to reduce the size of the embedded flash and then regenerates
+all auto-generated files. The flash size is reduced in a way that does not impact the register
+interface of the flash controller and is thus more or less transparent to software.
+"""
+
+import logging as log
+import subprocess
+import sys
+
+# Display INFO log messages and up.
+log.basicConfig(level=log.INFO, format="%(levelname)s: %(message)s")
+
+
+def main():
+
+    # Change the following strings in the following source files.
+    files = [
+        "hw/top_earlgrey/rtl/top_pkg.sv",
+        "hw/top_earlgrey/data/top_earlgrey.hjson"
+    ]
+    strings_old = [
+        "localparam int FLASH_PAGES_PER_BANK=256;",
+        "{ name: \"eflash\",\n" +
+        "      clock_srcs: {clk_i: \"main\"},\n" +
+        "      clock_group: \"infra\",\n" +
+        "      reset_connections: {rst_ni: \"lc\"},\n" +
+        "      type: \"eflash\",\n" +
+        "      base_addr: \"0x20000000\",\n" +
+        "      swaccess: \"ro\",\n" +
+        "      size: \"0x80000\""
+    ]
+    strings_new = [
+        "localparam int FLASH_PAGES_PER_BANK=32;",
+        "{ name: \"eflash\",\n" +
+        "      clock_srcs: {clk_i: \"main\"},\n" +
+        "      clock_group: \"infra\",\n" +
+        "      reset_connections: {rst_ni: \"lc\"},\n" +
+        "      type: \"eflash\",\n" +
+        "      base_addr: \"0x20000000\",\n" +
+        "      swaccess: \"ro\",\n" +
+        "      size: \"0x10000\""
+    ]
```

Comment:
I think this approach would make the script more complex as it will only work for the hjson but not for the SV file. Using regex for both of the files improves robustness already a lot. I would say that's a good middle ground.